src/porepy/utils/setmembership.py

```diff
@@ -179,9 +224,98 @@ def _numba_distance(mat, tol):
+    # practice failed.
     keep, new_2_old, old_2_new = _numba_distance(mat_t, tol)
 
     return mat[:, keep], new_2_old, old_2_new
+
+
+def uniquify_point_set(
+    points: np.ndarray[Any, np.dtype[np.float64]], tol: float = 1e-8
+) -> Tuple[
+    np.ndarray[Any, np.dtype[np.float64]],
+    np.ndarray[Any, np.dtype[np.int64]],
+    np.ndarray[Any, np.dtype[np.int64]],
+]:
+    """Uniquify a set of points so that no two sets of points are closer than a
+    distance tol from each other.
+
+    This function is partially overlapping by the function unique_columns_tol,
+    but the latter is more general, as it provides fast treatment of integer
+    arrays.
+
+    FIXME: It should be possible to unify the two implementations, however,
+    more experience is needed before doing so.
+
+    Parameters:
+        mat (np.ndarray, nd x n_pts): Columns to be uniquified.
+        tol (double, optional): Tolerance for when columns are considered equal.
+            Should be seen in connection with distance between the points in
+            the points (due to rounding errors). Defaults to 1e-8.
+
+    Returns:
+        np.ndarray: Unique columns.
+        new_2_old: Index of which points that are preserved
+        old_2_new: Index of the representation of old points in the reduced
+            list.
+
+    """
+    # The implementation uses Scipy's KDTree implementation to efficiently get
+    # the distance between points.
+    num_p = points.shape[1]
+    # Transpose needed to comply with KDTree.
+    tree = KDTree(points.T)
+
+    # Get all pairs of points closer than the tolerance.
+    pairs = tree.query_pairs(tol, output_type="ndarray")
+
+    if pairs.size == 0:
+        # No points were find, we can return
+        return points, np.arange(num_p), np.arange(num_p)
+
+    # Process information to arive at a unique point set. This is technical,
+    # since we need to deal with cases where more than two points coincide
+    # (example: if the points p1, p2 and p3 coincide, they will be identified
+    # either by the pairs {(i1, i2), (i1, i3)}, by {(i1, i2), (i2, i3)},
+    # or by {(i1, i3), (i2, i3)}).
+
+    # Sort the index pairs of identical points for simpler identification.
+    # NOTE: pairs, as returned by KDTree, is a num_pairs x 2 array, thus
+    # sorting the pairs should be along axis 1.
+    pair_arr = np.sort(pairs, axis=1)
+    # Sort the pairs along axis=1. The lexsort will make the sorting first
+    # according to pair_arr[:, 0] (the point with the lowest index in each
+    # pair), and then according to the second index (pair_arr[:, 1]). The
+    # result will be a lexiograpically ordered array.
+    # Also note the transport back to a 2 x num_pairs array.
+    sorted_arr = pair_arr[np.lexsort((pair_arr[:, 1], pair_arr[:, 0]))].T
+
+    # Find points that are both in the first and second row. This will identify
+    # triplets identified by pairs {(i1, i2), (i2, i3)} as described above.
```

Comment:
I guess I'm missing something here, but just in case: Quadruplets etc. are also caught, right?

Reply:
Yes, but fair point. I will improve documentation.